generated exception stubs

rschatz opened the pull request:
Use Graal-generated stubs instead of handwritten C stubs for throwing NullPointerException and ArrayIndexOutOfBoundsException.

rschatz commented:
/cc @tkrodriguez

tkrodriguez commented on graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/stubs/NullPointerExceptionStub.java:
```diff
@@ -0,0 +1,54 @@
+        if (index == 0) {
+            return providers.getRegisters().getThreadRegister();
+        } else {
+            return null;
```
this case should probably be an assert or an explicit throw

dougxc replied:
Yes, the `index` should be asserted to be 0 before its use.

christhalinger replied:
Why exactly?  The other stubs have a switch-case and don't assert in the default case.

dougxc replied:
Depends on which other stubs are you looking at. Both `ExceptionHandlerStub.getConstantParameterValue` and `UnwindExceptionToCallerStub.getConstantParameterValue` use assertions. These however should be `JVMCIError.guarantee` - there's no performance cost to unconditionally doing the check.

christhalinger replied:
True.  I left a comment in `OutOfBoundsExceptionStub`.

tkrodriguez replied:
These might make more sense if they were just @Fold methods.  Usually parameters are used to provide different variations but that's not happening here.  These are all singletons.

rschatz replied:
These `@Fold` methods would not be static, so we need to get something in there, either the `HotSpotProviders`, or the `this` object. And when we're passing in something, we might as well directly pass the `Register` object.

For the others (in OutOfBoundsExceptionStub), i agree, an `@Fold` method might be better.

tkrodriguez commented:
After seeing all the changes required on both sides and the code required to construct the right messages I'm wondering if we should just be providing the same entry points that C1 has for throwing implicit exceptions in c1_Runtime1.cpp from JVMCIRuntime.  The set of exceptions that have to be handled in this way is limited I think since they are only required for exceptions which are thrown implicitly.  I guess we could have a case where we wanted stubs for some other case.

tkrodriguez commented on graal/com.oracle.graal.hotspot/src/com/oracle/graal/hotspot/stubs/OutOfBoundsExceptionStub.java:
```diff
@@ -0,0 +1,86 @@
+            ptr.writeByte(0, (byte) '-');
+        }
+
+        return createException(threadRegister, ArrayIndexOutOfBoundsException.class, ptr);
```
Don't you need to null terminate the string?

dougxc replied:
He's building the string backwards and starts with the null terminator.

rschatz commented:
re "just export all exception throwing routines C1 has available":
We would be introducing a lot of code that's never exercised because Graal doesn't use it yet. And on the other hand, we'd need to be very sure that we don't miss anything. The solution with generating the stubs ourselves is more future-proof, we can always just create another stub.

rschatz commented:
New version addressing the comments raised so far:
- wrong index in `getConstantParameterValue` now throws an exception
- removed `bufferSize` constant parameter, since it doesn't need access to `providers` (the suggestion was to use an `@Fold` method, but we actually don't need even that)

I've also included a `SymbolPointer` word-type, analogous to the existing `KlassPointer` and `MethodPointer` types.

What I haven't addressed yet is that we need variants of `throw_and_post_jvmti_exception` for exception messages that contain a `Klass::internal_name()` (`ArrayStoreException` and `ClassCastException`). Since those two are currently not used in Graal, I'd like to merge this PR now, and do a separate PR for supporting the others.

tkrodriguez commented:
I meant to comment last night that it looks good to me, but forgot to, so just for the record, lgtm.
